Week 3 code

hieudan225 opened the pull request:
Fetch data from server to change the background image of index.html

hieudan225 committed:
Update the README.md file to describe the repo.

hieudan225 committed:
- update the skeleton content and visual of two pages ('home' and 'about')
- add 'swup' module used for transition animation between pages

hieudan225 committed:
Clean up the previous version by reformating the html and css files.

hieudan225 committed:
Relocate the node_modules folder to webapp folder.

hieudan225 committed:
Delete Swup module (not working)
Finish the skeleton webpages without any animation yet

hieudan225 committed:
combine the css configuration of index.html and intro.html, retry swup module

hieudan225 committed:
test servlet (step1-week3)

hieudan225 committed:
Fetch an imageURL from server to change the background-image of index.html

hieudan225 committed:
- Fetch a JSON object (ArrayList<String> images) from server
- Display the array in the comment section of intro.html

hieudan225 committed:
Handle doPost, doGet functions for the new comment sections (Week 3 - Step 4)

cthamrun reviewed (approved):
Changes look good to me. Feel free to add Grace as reviewer as well so you can get more feedback on your code :)

hieudan225 committed:
Store comments data using datastore

hieudan225 committed:
Fix onload bug.
Before, when intro.html was loaded from index.html, the comment history was not showing up. This was due to the Swup module loading function.
Now, Swup has been removed and the comment history is showing when the page loads.

ganewman42 commented:
Could you include a screenshot of your updated website? Or is the deployed version up to date?

P.S. I really like the design language, your choice of fonts and colors are very engaging.

ganewman42 commented on README.md:
```diff
@@ -1,13 +1,13 @@
   - If the tutorial panel does not open or display the walkthrough contents, try refreshing the page.
-
 Then follow the on-screen instructions to set up your repo.
+dhle@cloudshell:~/step$ nano README.md
```
I think this might be an accidental change, please remove. If they're purposeful instructions they could be a little more clear.

ganewman42 commented on portfolio/package-lock.json:
```diff
@@ -0,0 +1,3 @@
+{
```
Adding a package lockfile is a super smart move! Way to use great tooling here.

ganewman42 commented on portfolio/src/main/java/com/google/sps/data/Comment.java:
```diff
@@ -0,0 +1,17 @@
+        this.id = id;
+        this.name = name;
+        this.content = content;
+        this.timestamp = timestamp;
```
Optional: I think Java has a native time time, LocalDateTime. Using a native time type could make dealing with timezones and type conversions a little easier.

hieudan225 replied:
I totally agree. Let me change that.

ganewman42 commented on portfolio/src/main/java/com/google/sps/servlets/BackgroundChange.java:
```diff
@@ -0,0 +1,35 @@
+  private int prevIndex = -1;
+  @Override
+  public void init() {
+      images = new ArrayList<>();
```
Every time you add a new image for the random selector, you'll have to remember to change this list. If you instead use a mechanism like getting the filenames from this directory and iterating over them, simply adding an image to the "image" directory or a new subdirectory for backgrounds would be sufficient. Not having to make a change in two places makes your code much easier to maintain.

hieudan225 replied:
That is a great suggestion! I'll change that.

ganewman42 commented on portfolio/src/main/java/com/google/sps/servlets/BackgroundChange.java:
```diff
@@ -0,0 +1,35 @@
+public class BackgroundChange extends HttpServlet {
+  private List<String> images;
+  private int prevIndex = -1;
+  @Override
```
Nit: It's nice to have an extra newline between function declarations. Please add these throughout.

hieudan225 replied:
I have changed that.

ganewman42 commented on portfolio/src/main/java/com/google/sps/servlets/DeleteComment.java:
```diff
@@ -0,0 +1,34 @@
+
+/** Servlet that returns a list of comments */
+@WebServlet("/delete-comment")
+public class DeleteComment extends HttpServlet {
```
Deletions are a smart addition.

hieudan225 commented:
Hi @ganewman42, I think this is the link of my updated website: https://dhle-step-2020.ue.r.appspot.com. Here is 2 screenshots of my new features (background-change using a picture from the server and the comment section): 

![image](https://user-images.githubusercontent.com/42331170/83441092-f598ff00-a413-11ea-8eda-8cf1f2dd9cb4.png)

![image](https://user-images.githubusercontent.com/42331170/83441046-e1550200-a413-11ea-9243-b38f9d311f0b.png)

ganewman42 commented on portfolio/src/main/java/com/google/sps/servlets/ListComments.java:
```diff
@@ -0,0 +1,63 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
```
So a few things on licenses:

* Just because you use Java does not mean YOUR code needs to have the Apache license
* Depending on how open source you want your code to be, some flavor of a gnu public license might be more appropriate
* Generally you put your license once in the readme instead of in each code header (otherwise it clutters things up to have a long header in a bunch of places)

Let me know if you have questions about licensing and that whole world.

hieudan225 replied:
I see. Those line of licenses documentation was in the file originally, so I didn't know that it should be deleted. I have deleted those licenses.

ganewman42 commented on portfolio/src/main/java/com/google/sps/servlets/ListComments.java:
```diff
@@ -0,0 +1,63 @@
+    List<Comment> comments = new ArrayList<>();
+    for (Entity entity: results.asIterable()) {
+        long id = entity.getKey().getId();
+        String name = (String) entity.getProperty("name");
```
Though experiment: What about little bobby drop tables? Do you need to sanitize this data? (Context: https://xkcd.com/327/)

hieudan225 replied:
Oh I didn't know about sanitizing data. I'll surely spend time doing that today!

ganewman42 commented on portfolio/src/main/webapp/script.js:
```diff
@@ -1,28 +1,44 @@
+        let content = document.createElement("span");
+        content.innerText = "Content: " + comment.content;
+        
+        let deleteElement = document.createElement("button");
```
Nit: I notice that this button is right up against the end of the comment, you might want to add some additional spacing here.

hieudan225 replied:
This is really weird. I don't think I have getRandomGreeting() function in my script.js file any more.

ganewman42 commented on portfolio/src/main/webapp/index.html:
```diff
@@ -1,18 +1,56 @@
+                <button onclick="getRandomImageAsync()"
+                        style="position: absolute;
+                        top: 90%;
+                        left:25%;">Click here to change the background with a random image from the server</button>
```
Nit: A shorter button might be good, something like "Click here to change the background!" "Different Background" might be nice.

ganewman42 reviewed (commented):
Looks great! As I said it's a very nice looking website where you use some very clever things.  Comments are mostly cleanups, optional items, nitpicks (nits), and things to think about are noted where applicable. Let me know if you have questions.

hieudan225 committed:
Remove accidental change in README.md

hieudan225 committed:
Remove accidental change in README.md

ganewman42 reviewed (approved):
LGTM in case you were waiting for my review
